charts/kube-prometheus-stack/crds/crd-prometheuses.yaml

```diff
@@ -5,6 +5,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
+    argocd.argoproj.io/sync-options: Replace=true
```

Comment:
i dont think we should add deployment specific annotations to the crd.
imho the change in the docs for manual upgrade should be enough.

Reply:
@monotek The problem here, is that CRDs are not templated. There is no way to add annotations except hacky workarounds.

Reply:
@sathieu Great idea to add this directly to the CRD; it solves the problem for ArgoCD users and the annotation will simply be ignored for everyone else.  The "manual upgrade" isn't really a viable option when using ArgoCD; every sync attempt will fail because the CRD is too large.  Without this change, the only way for an ArgoCD user to fix this is to replace *everything* on each sync operation (risky) or to extract the CRD from the chart so it can be updated locally.

Reply:
As helm is not supposed to update crds anyway, most people add and manage these crds "by hand" or via some other automated step, outside of helm..
Therefore the crd does not need such annotations. 
In the long term this would make the CRD also kind of messy, if everybody starts to add things for his special deployment method, which in the end are not realy needed.

Reply:
Guys, we have to pay attention to the way the CRDs are updated in this chart. It's not thru manual changes like the above but thru an automated process that pulls them from the Prometheus Operator repo. So this change here is even pointless because it will get overwritten upon the next major update of the CRDs.

Reply:
@Xtigyro Wont it be added because of the changed update_crd script?

Reply:
@steinarox Ah, apologies - didn't realize we were updating that file after every pull from the Prometheus Operator repo. You are right.

Reply:
And one of the reasons I didn't realize that was because that script was not meant to modify the CRDs.

Reply:
Maybe the other maintainers might have another take on the matter but I'm with @monotek on this one. The chart is not meant to create new opinionated versions of the used external code. 

Having said that - @sathieu how about adding this functionality in a completely optional way. That is to say - adding another section in the `README.md` about tools like ArgoCD, Flux, etc. - and state what particular changes are needed to make the CRDs fully compatible with third-party tools that operate in a non-standard way. Maybe even putting another small script that does the changes upon manual execution?

Reply:
@Xtigyro There is no easy way to workaround this on the ArgoCD side.

Possible correct solutions:

- Implement server-side apply in ArgoCD https://github.com/argoproj/argo-cd/issues/2267
- Support helm post renderers in ArgoCD https://github.com/argoproj/argo-cd/issues/3698

Heavy workaround:

- https://github.com/argoproj/argocd-example-apps/issues/103#issuecomment-990793051

I won't work more on this as I have a workaround (thanks to Ansible, see https://gitlab.com/kubitus-project/kubitus-installer/-/issues/70). But this workaround is clearly not easily actionable to everyone.

Reply:
> As helm is not supposed to update crds anyway, most people add and manage these crds "by hand" or via some other automated step, outside of helm..

What examples exist that this is how "most people" managed CRDs? The kube-prometheus-stack Helm chart includes the CRDs and applying/creating CRDs through Helm/ArgoCD is common practice and works.

> Great idea to add this directly to the CRD; it solves the problem for ArgoCD users and the annotation will simply be ignored for everyone else.

Agreed; this is a simple and harmless 1-line annotation that solves what is otherwise a huge pain for administrators, and does not affect those not using ArgoCD. Is it really worth it to take a philosophical stance against "add[ing] deployment specific annotations to the crd"?